### PR TITLE
Reduce performance footprint of Pumpkin Warden AI

### PR DESCRIPTION
**Pull request by Geolykt**
Most of the "magic" behind this commit is that there is now a delay between polling the goals of the pumpkin wardens.

The impact is still comparatively high (~3-4% of the tick) but it's far better than the >50% of the tick (~25 mspt) I have seen before (which prompted me to make this PR)

I guess it would be possible to optimize it further by implementing some kind of a subscriber system where once a pumpkin grows a warden gets notified and takes appropriate action. However in vast majority of cases that is a bit overkill. Especially since the infrastructure needed for such a system would need to be present in memory regardless of the Pumpkin wardens being loaded or not.

That being said there might be a more intelligent approach I might have missed due to me being rather unfamiliar with fabric and forge.

On that regard I also tried to set the return value of `nextStartTick` to a non-0 value, but it seems that induces significant amounts of jank (most notably there being a visible delay until a warden break one of the blocks). The solution I have gone with is also a bit flawed but I cannot really find the source of it and is most likely induced by other factors (perhaps other goals interfering?).

---

The profiling was done on a ATM 8 install with a CPU that definitely should not be able to run that modpack - so the values are a bit skewed. But hey - it's representative of real-world usage.

Performance before: https://spark.lucko.me/2BJ2xjX1Ei
Performance after: https://spark.lucko.me/zni2PJoJoA

### CODE REVIEW

**Comment by JT122406** [comment]
Can you run the same tests but with just BYG and Terrablender along with spark on both Forge and Fabric?
Just will give us some better ideas on where the performance sits

**Comment by JT122406** [comment]
Mainly asking because the TPS difference between the 2 reports is so significant that the improvement could be way exaggerated.

**Comment by Geolykt** [comment]
I'll look at it tomorrow. The improvements will be exaggerated either way due to my being on a rather budget CPU though

File: Common/src/main/java/potionstudios/byg/common/entity/pumpkinwarden/PumpkinWarden.java [hunk 1]
```diff
@@ -254,19 +254,50 @@ public BlockState getCarriedBlock() {
...
                 return false;
             }
             return super.canUse();
         }
 
+        @Override
+        protected boolean findNearestBlock() {
+            if (this.searchNearestBlockDelay == -1 && this.isValidTarget(this.warden.level, this.blockPos)) {
+                return true;
+            } else if (this.searchNearestBlockDelay > 0) {
```
Comment (JT122406) [inline Common/src/main/java/potionstudios/byg/common/entity/pumpkinwarden/PumpkinWarden.java]: Is the number this.searchNearestBlockDelay more likely going to be -1 or any number > 0?

Comment (Geolykt) [reply Common/src/main/java/potionstudios/byg/common/entity/pumpkinwarden/PumpkinWarden.java]: Even if `searchNearestBlockDelay` is most likely to be `-1` the comparision could be performed concurrently without possibly changing the program outcome, so I actually expect it to be more performant if we rearrange the blocks regardless of likely valid (though your suspicion that it is more likely to be > 0 is correct).

Comment (JT122406) [reply Common/src/main/java/potionstudios/byg/common/entity/pumpkinwarden/PumpkinWarden.java]: Yeah very small but better if we just swap the blocks.

**Comment by Geolykt** [comment]
Before (Forge): https://spark.lucko.me/ioyqixeAu7
Before (Fabric): https://spark.lucko.me/iIU7k4RX79
After (Forge): https://spark.lucko.me/AV9U4kKYCl
After (Fabric): https://spark.lucko.me/3lWx3pP6hL

What a waste of my time.

Before you come shouting at me that the improvements are smaller than advertised: That is because you are looking at relative values (% of tick). The absolute values (ms per tick) are much more drastic. Regardless I could reduce the numbers even more by increasing the delay to more drastic numbers but then I am pretty sure this PR won't even get merged because the prior lack of a delay and thus the inevitable lag is working as intended in the first place

**Comment by JT122406** [comment]
Actually the change is huge on forge 
image

**Review by JT122406 — APPROVED** [review]

**Pull request merged**